5_Stable_implementation/flip.cpp

```diff
@@ -0,0 +1,332 @@
+  auto start = std::find_if(tl.begin(), tl.end(), isDefined);
+  auto stop = std::find_if(tl.rbegin(), tl.rend(), isDefined);
+  auto it = std::find_if(start, stop.base(), isNull);
+  return it == stop.base();
+}
+
+// Transposes the tensor and indices together so that all the non-null indices
+// index the first k dimensions of the tensor. Returns the transposed tensor
+// and the reordered indices. For example:
+// transposeToFront(tensor, {nullptr, a, nullptr, b})
+// returns
+// tensor.permute([1, 3, 0, 2]), {a, b, nullptr, nullptr}
+static std::tuple<torch::Tensor, std::vector<torch::Tensor>, std::vector<int64_t>>
+transposeToFront(torch::Tensor self, torch::TensorList indices) {
+  std::vector<int64_t> dims;
+  std::vector<int64_t> inv_dims(self.dim());
+  std::vector<torch::Tensor> transposedIndices;
+
+  dims.reserve(self.dim());
+  for (auto i = decltype(self.dim()){0}; i < self.dim(); i++) {
+    if (indices[i].defined()) {
+      inv_dims[i] = dims.size();
+      dims.push_back(i);
+      transposedIndices.emplace_back(indices[i]);
+    }
+  }
+  for (auto i = decltype(self.dim()){0}; i < self.dim(); i++) {
+    if (!indices[i].defined()) {
+      inv_dims[i] = dims.size();
+      dims.push_back(i);
+      transposedIndices.emplace_back();
+    }
+  }
+  auto permuted_input = self.permute(dims);
+
+  // std::reverse(dims.begin(), dims.end());
+  return std::make_tuple(permuted_input, std::move(transposedIndices), inv_dims);
+}
+
+// Replace indexed dimensions in src with stride 0 and the size of the result tensor.
+// The offset in these dimensions is computed by the kernel using the index tensor's
+// values and the stride of src. The new shape is not meaningful. It's used to make
+// the shape compatible with the result tensor.
+static torch::Tensor restride_src(const torch::Tensor& src, int64_t dims_before, int64_t dims_indexed,
+                                  torch::IntArrayRef replacement_shape) {
+  auto shape = torch::DimVector(src.sizes());
+  auto strides = torch::DimVector(src.strides());
+  int64_t end = dims_before + dims_indexed;
+  shape.erase(shape.begin() + dims_before, shape.begin() + end);
+  strides.erase(strides.begin() + dims_before, strides.begin() + end);
+  shape.insert(shape.begin() + dims_before, replacement_shape.begin(), replacement_shape.end());
+  strides.insert(strides.begin() + dims_before, replacement_shape.size(), 0);
+  return src.as_strided(shape, strides);
+}
+
+// Add dimensions of size 1 to an index tensor so that it can be broadcast to the result
+// shape and iterated over element-wise like the result tensor and the restrided src.
+static torch::Tensor reshape_indexer(const torch::Tensor& index, int64_t dims_before, int64_t dims_after) {
+  auto orig_shape = index.sizes();
+  auto shape = torch::DimVector();
+  shape.append(dims_before, 1);
+  shape.append(orig_shape.begin(), orig_shape.end());
+  shape.append(dims_after, 1);
+  return index.reshape(shape);
```

Comment:
I think this can be simplified as well, like @vfdev-5 did for interpolate, where we keep all dimensions as 1 and only the dimension that is indexed that we consider to have the shape of the dim.